### ISSUE TITLE
Bluesky: support account migration

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#1137]
Specifically, let people migrate an existing Bluesky account into BF to become a bridged account, or an existing bridged Bluesky account out of BF to become a normal Bluesky account that they can log into and use.

Big project! AP corollary is #330. Background:

* https://atproto.com/guides/overview#account-portability
* https://docs.bsky.app/blog/self-host-federation#account-migration

**Comment by tesaguri** [comment]
Is the plan to allow using an atproto data repository as a bridged account and a "normal" account at the same time, or is it going to be mutually-exclusive, that is, you won't be able to manually write your own atproto records into the data repository while it's used as a bridged account?

**Comment by snarfed** [comment]
There isn't a plan for this yet, it's honestly pretty unlikely, but if it does happen, it'd be the latter. I definitely don't plan to support bridged accounts that can also be used "manually" like normal accounts.

**Comment by snarfed** [comment]
Now that `goat` can migrate Bluesky accounts, this might already be possible. I'd love to hear how it goes if anyone wants to try!

Sadly the "easy" way requires logging into the source PDS, which Bridgy Fed doesn't support, but the "manual" way doesn't: https://whtwnd.com/bnewbold.net/3l5ii332pf32u

**Comment by rossabaker** [comment]
I haven't tried it, but I'm studying the fire escapes.  The part of the manual process that looks most troublesome to me is:

> Retrieve the token ($PLCTOKEN) from email, then request a signed version of the PLC params:
>
> ```sh
> # old PDS
> goat account plc sign --token $PLCTOKEN  ./plc_unsigned.json > plc_signed.json
> ```

I don't think BF has our email?  Would this step need to be exposed as a command to the bot, both to avoid the login and to get the token delivered?

**Comment by Tamschi** [comment]
> […] Would this step need to be exposed as a command to the bot, both to avoid the login and to get the token delivered?

Yes, I think that's the case. I'm not sure this is feasible to do via DM, since most software doesn't let you attach arbitrary files there.
Maybe if fetching the "recommended" DID parameters from the new PDS doesn't require authentication, then Bridgy fed could get that file directly.

Alternatively, I think Bridgy Fed could expose a recovery key to the user for them to sign the PLC operation offline.
I'm really not sure how to look for instructions on that, though, and how it integrates in this process here.

**Comment by snarfed** [comment]
Yes! Thanks, you're both right.

For now, if anyone wants to do all of the manual migration steps up to the PLC DID doc update, I'll happily do that update manually. After that, if there's enough demand for this, I can implement the methods that `goat` needs to do it automatically.

**Comment by Tamschi** [comment]
cc @alien-sunset I'm not sure if you're subscribed here, but would that work for you?
This way you could take control of the current bridge account and post to it directly when botsin.space shuts down.

**Comment by alien-sunset** [comment]
Thanks, but I’ve already set up a new account and coded the bot to post directly to Bsky (which took way too long, I am NOT a good coder 😅)

**Comment by Daft-Freak** [comment]
IDK if I'm missing something, but the manual method seems to also require logging into the source PDS pretty much immediately.

**Comment by snarfed** [comment]
@Daft-Freak try logging into the destination PDS instead? 

The bsky.social PDS doesn't yet support migrating accounts into it, and I haven't set up a separate PDS myself to test, but it's possible that it could work!

**Comment by Daft-Freak** [comment]
Yeah, it definitely needs a token from the source PDS, generating one from an account on the dest results in: `error: failed to create account: XRPC ERROR 401: AuthenticationRequired: Missing auth to create account with did: [...]`

(I'm not in any rush to actually do this, mostly just curious about the possibility)

**Comment by snarfed** [comment]
@Daft-Freak just to confirm, this is on a self-hosted PDS, not the main bsky.social one, right?

bsky.social doesn't allow inbound migration right now, but supposedly self-hosted PDSes should...

**Comment by Daft-Freak** [comment]
Yep, self-hosted PDS.

**Comment by snarfed** [comment]
Interesting! cc @bnewbold, I doubt that's expected. Is there maybe a setting you need to enable on the PDS to allow creating new accounts as `deactivated`?

Assuming this error is on the `goat account create` command, I don't think there's anything I can do about this on Bridgy Fed's side, that command is going to your PDS. It needs to allow creating new accounts.

Maybe paste in all of the commands you ran, including eg `goat account login` and `goat account service-auth`, without passwords and tokens, and we can look?

**Comment by bnewbold** [comment]
The "smooth" account migration path is described here, basically documenting what `goat` does internally:

https://github.com/bluesky-social/atproto/discussions/3176

It requires a number of endpoints implemented on the "old" PDS:

- `com.atproto.server.getServiceAuth`: used in first steps so that client (eg, `goat`) can "prove" to new PDS that it is the account in question. if the old PDS doesn't implement or is offline, the client needs control of the identity (DID document) and could generate such a proof locally (using the atproto signing key in the current DID document).
- `com.atproto.server.checkAccountStatus`: probably not really required on the old PDS, but is mentioned in migration guide
- `com.atproto.sync.getRepo`: for CAR export
- `com.atproto.sync.listBlobs`: enumerates all the blobs; client will pull from this list, then download each with `getBlob`
- `com.atproto.identity.requestPlcOperationSignature`: for DID PLC, this is a request to generate a signed PLC op (using old PDS rotation key), for the case where client doesn't have control of the identity (DID document). sometimes this requires a secret token, eg delivered via account email.
- `com.atproto.server.deactivateAccount`: as a final step on old PDS

All of these are authenticated methods, so also require login and session endpoints.

I think bridgy pseudo-PDS does not support a bunch of these, and maybe doesn't make sense to? Eg, password login/auth.

Some of the full-offline mechanisms (like self-signing a "service auth token") are important for the unavailable-old-PDS flow, but are not yet implemented in `goat`. This is a good reminder to get around to implementing those and documenting that migration process.

**Comment by bnewbold** [comment]
I created an issue to track the later functionality in `goat` here: https://github.com/bluesky-social/indigo/issues/864

**Comment by snarfed** [comment]
When migrating a bridged account out to be a native Bluesky account, one nice bit of polish would be to set the bridged handle's DNS (eg snarfed.org.web.brid.gy) to the native account's DID, which will make it redirect there in some clients and situations. Not an officially supported ATProto/Bluesky feature afaik, but it does work, kind of.

**Comment by snarfed** [comment]
For everyone following this issue, we finally launched https://bounce.anew.social/ in beta! https://blog.anew.social/bounce-beta-now-live/

It currently supports migrating from Bluesky to fediverse, while keeping all of your followers and many of your follows. We're actively working on adding the other direction too, fediverse to Bluesky.

**Comment by lupomancer** [comment]
Re: the migration here, I still feel like there's a bit of risk with the potential of just losing your account/instance despite everyone's best efforts. 

Any ideas on how that could be rectified? In my head, I've kinda thought of allowing your bridgy account to be, like, re-linked to a new mastodon account in the case of your instance going down or something. With the BSKY PDS, you're also relying on them not going down, but I think that's far less likely than your mastodon instance randomly going dark for some reason, in which case you'd lose both sides of that network. Or, even with notice, if your instance shuts down there's no real way to migrate those BSKY followers over to the new account even if the Fedi side will migrate all your fedi followers. 

Unless I'm mistaken! I could be misinterpreting the documentation for Bridgy. 

I'm also not necessarily looking directly for an answer immediately, I'm more just curious if this is on your radar/in scope. With the DM for unabridged replies feature, it's becoming more and more attractive to use my bridged mastodon as my main and my BSKY as a reply account, but these little corner cases kinda spook me

**Comment by snarfed** [comment]
@lupomancer yes! Absolutely right. If your fediverse instance disappears, it'd be nice to be able to attach its clone Bluesky account to a new fediverse account, or migrate it out to be a native Bluesky account.

The main problem is, how would we authenticate you? All we really know is your fediverse account, which is now inaccessible.

(If your fediverse instance does an orderly shutdown, with time for you to export/migrate out while it's still up, the proposed feature here will work for you as is.)

**Comment by lupomancer** [comment]
The authentication question is...one I didn't think about, hahaha!

I mean the EASY (to say in a comment) answer would be to have some kind of local/bridgy based auth, but then you would need to build out a system for all of that which seems like too much work at this point.

Maybe, for people using custom domain handles, there could be a DNS record entry to prove ownership? But that would only prove that the domain is connected to the account, and not that the user currently on the bridgy control panel is the owner.

Yeah, I don't know, I don't really have an answer for that!

As for the current system's handling of account migration -- I remember reading in the docs (a while ago) that migrating to a new fedi account basically meant that the bridged BSKY account would have to be started over new. Does Bridgy currently handle migrations like that? Or would it connect the BSKY-side account to the new migrated-to fedi account? Again, I could totally be misunderstanding the docs and comments on other issues so please correct me if I'm wrong!

**Comment by snarfed** [comment]
Hmm, Bridgy Fed doesn't really do migration at all right now. Do you mean Bounce's handling?

**Comment by snarfed** [comment]
Oh, if you have a native fediverse account, and bridge it, and then migrate it to a different native fediverse account? Then yes, you'd have to re-bridge it, and it would start over with a separate new Bluesky clone account. We can do better eventually! This issue would cover that.

**Comment by ppp-x** [comment]
Here's my situation: I have a Pixelfed account, and I have a Bluesky account. I want to migrate the Bluesky account to either Bridgy Fed or Bounce, and have my existing Pixelfed account bridged to that Bridgy Fed/Bounce atproto account (so basically let my existing Bluesky account continue as a bridged Pixelfed account). Is that a thing?

I should add that my Pixelfed account already has a bridge to Bluesky but that bridge can be purged as it has hardly any followers.

**Comment by snarfed** [comment]
@ppp-x sorry, no, we don't bridge existing native accounts to each other. You'll probably want to look at cross-posting clients or similar setups instead.

**Comment by ppp-x** [comment]
> You'll probably want to look at cross-posting clients or similar setups instead.

No, I had actually hoped for a way of importing my Bluesky account to Bounce (thus discontinuing it as a native Bluesky account), and bridge my Pixelfed account to it. But alas.

**Comment by snarfed** [comment]
Ah! I misunderstood. That's exactly what Bounce does!

**Comment by Donad678** [comment]
> [@lupomancer](https://github.com/lupomancer) yes! Absolutely right. If your fediverse instance disappears, it'd be nice to be able to attach its clone Bluesky account to a new fediverse account, or migrate it out to be a native Bluesky account.
> 
> The main problem is, how would we authenticate you? All we really know is your fediverse account, which is now inaccessible.
> 
> (If your fediverse instance does an orderly shutdown, with time for you to export/migrate out while it's still up, the proposed feature here will work for you as is.)

I would love to have a way to transfer the bsky account to a different fedi account. My issue is that my old instance is gone, files broken and backups corrupted gone. And now I can't link it to my new account. I even had a custom domain set, which would probably one way to authenticate actually, as you could for example let the user set a txt record to check or something like that. But that would only work for people that used a custom domain in the first place

**Comment by FiloSottile** [comment]
A minimal implementation of this would be to let users add a PLC key. Then they could use any of the previous-PDS-is-offline migration tools, which have gotten a bit better over the past year (or worse come to worse, keep their followers by creating a new account for the DID without migrating the data).

**Comment by snarfed** [comment]
Done! https://blog.anew.social/migrations-update-launch/
